### ISSUE TITLE
Match statement within loop complains about non exhaustiveness

### ISSUE THREAD

**Issue by rslinckx** [microsoft/pyright#4526]
**Describe the bug**

pyright 1.1.286 -> 1.1.291 (included) have a problem with exhaustiveness check in match statement, when the statement is inside a loop

pyright 1.1.285 works fine

Error message:
```
Cases within match statement do not exhaustively handle all values
  Unhandled type: "Unknown"
  If exhaustive handling is not intended, add "case _: pass"
```

**To Reproduce**
```
import enum

@enum.unique
class MyEnum(enum.Enum):
    A = enum.auto()
    B= enum.auto()

def test_with_loop(source: MyEnum | None = None) -> None:
    for _ in []:
        match source: # <- Here pyright will complain, reveal_type() will show MyEnum|Unknown|None
            case None:
                pass
            case MyEnum.A:
                pass
            case MyEnum.B:
                pass
            
def test_without_lopp(source: MyEnum | None = None) -> None:
    match source: # <- Here everything's fine
        case None:
            pass
        case MyEnum.A:
            pass
        case MyEnum.B:
            pass

```

I'm not sure why there is a Unknown type inference inside the loop.

**Comment by erictraut** [comment]
I'm not able to repro the problem with the code above. Could you confirm that you are able to repro it with a clean project and with default configuration settings?

**Comment by rslinckx** [comment]
Sorry, I think I might have mixed up my various versions between what's provided in pylance and pyright itself.

So pyright 1.1.289 has the problem, and indeed 1.1.290 and 1.1.291 seem to be working with the above example.

The problem however seems to be a little bit more tricky:
```
import enum


@enum.unique
class MyEnum(enum.Enum):
    A = enum.auto()
    B = enum.auto()


def test_fail(source: MyEnum | None = None) -> list[str]:
    results = []
    for x in []:
        match source: # <- Complains
            case None:
                results.append(x)
            case MyEnum.A:
                pass
            case MyEnum.B:
                pass
    return results


def test_success(source: MyEnum | None = None) -> list[str]:
    results = []
    for x in []:
        match source: # <- No complain. Note the print(x) instead of results.append(x)
            case None:
                print(x)
            case MyEnum.A:
                pass
            case MyEnum.B:
                pass
    return results
```

**Comment by rslinckx** [comment]
My config file is as follows:
```
{
	"useLibraryCodeForTypes": true,
	"reportDuplicateImport": true,
	"reportUntypedNamedTuple": true,
	"reportUnnecessaryIsInstance": true,
	"reportUnnecessaryCast": true,
	"reportUnnecessaryComparison": true,
	"reportUnnecessaryTypeIgnoreComment": true,
	"reportImplicitStringConcatenation": true,
	"reportMatchNotExhaustive": true,
	"reportWildcardImportFromLibrary": true,
	"reportConstantRedefinition": true,
	"reportOverlappingOverload": true,
	"reportUnusedClass": true,
	"reportInvalidStubStatement": true,
	"reportIncompleteStub": true,
	"reportUnusedExpression": true,
	"reportCallInDefaultInitializer": true,
	"strictListInference": true,
	"strictSetInference": true,
	"reportMissingParameterType": true,
	"reportUntypedClassDecorator": true,
	"reportUntypedBaseClass": true,
	"reportUnusedFunction": true,
	"reportUntypedFunctionDecorator": true,
	"reportUnusedVariable": true,
	"reportInconsistentConstructor": true,
	"reportMissingTypeArgument": true,
	"reportUnusedImport": true,
	"reportUnnecessaryContains": true,
	"reportShadowedImport": true,
	"reportSelfClsParameterName": true,
	"reportFunctionMemberAccess": true,
	"reportUnknownParameterType": true,
	"strictDictionaryInference": false,
	"reportPrivateUsage": false,
	"reportPropertyTypeMismatch": false,
	"reportImportCycles": false,
	"reportIncompatibleMethodOverride": false,
	"reportIncompatibleVariableOverride": false,
	"reportMissingSuperCall": false,
	"reportUninitializedInstanceVariable": false,
	"reportUnknownArgumentType": false,
	"reportUnknownLambdaType": false,
	"reportUnknownVariableType": false,
	"reportUnknownMemberType": false
}
```

Running like: `npx pyright@1.1.291 -p pyrightconfig.ci.json enumfail.py`

**Comment by erictraut** [comment]
Thanks for the repro. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.292, which I just published. It will also be included in a future release of pylance.
